### PR TITLE
users: show information about groups

### PR DESCRIPTION
**Pull request by KKoukiou**

### CODE REVIEW

File: pkg/users/accounts-list.js [hunk 1]
```diff
@@ -38,11 +39,51 @@ import { SearchIcon } from '@patternfly/react-icons';
...
+        else
+            return b.members - a.members;
+    });
+};
+
+const GroupActions = ({ group, accounts }) => {
+    const [isKebabOpen, setKebabOpen] = useState(false);
+
+    if (!superuser.allowed)
+        return null;
```
Comment (cockpituous) [inline pkg/users/accounts-list.js:67]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221202-154936-cbc812c3-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/accounts-list.js [hunk 2]
```diff
@@ -38,11 +39,51 @@
...
+
+    const actions = [
+        <DropdownItem key="delete-group"
+                      className={group.uid === 0 ? "" : "delete-resource-red"}
```
Comment (cockpituous) [inline pkg/users/accounts-list.js:71]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221202-154936-cbc812c3-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/accounts-list.js [hunk 3]
```diff
@@ -183,17 +249,77 @@
...
+            const bname = b.columns[0];
 
-    const filtered_accounts = accounts.filter(account => {
-        if ((account.uid < 1000 && account.uid !== 0) ||
-                 account.shell.match(/^(\/usr)?\/sbin\/nologin/) ||
-                 account.shell === '/bin/false')
-            return false;
+            // current user is always first
+            if (admins.includes(aname.title))
+                return direction === SortByDirection.asc ? -1 : 1;
```
Comment (cockpituous) [inline pkg/users/accounts-list.js:270]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221202-154936-cbc812c3-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/accounts-list.js [hunk 4]
```diff
@@ -183,17 +249,77 @@
...
+            if (admins.includes(bname.title))
+                return direction === SortByDirection.asc ? 1 : -1;
```
Comment (cockpituous) [inline pkg/users/accounts-list.js:272]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221202-154936-cbc812c3-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/delete-group-dialog.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                }
+            ]
+        };
+
+        if (!dlg)
+            dlg = show_modal_dialog(props, footer);
+        else {
+            dlg.setProps(props);
+            dlg.setFooterProps(footer);
+        }
```
Comment (cockpituous) [inline pkg/users/delete-group-dialog.js:71]: These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221202-154936-cbc812c3-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/accounts-list.js (hunk 1 above)
Comment (cockpituous) [inline pkg/users/accounts-list.js:67]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221203-084140-31427c73-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/accounts-list.js (hunk 2 above)
Comment (cockpituous) [inline pkg/users/accounts-list.js:71]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221203-084140-31427c73-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/accounts-list.js (hunk 3 above)
Comment (cockpituous) [inline pkg/users/accounts-list.js:270]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221203-084140-31427c73-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/accounts-list.js (hunk 4 above)
Comment (cockpituous) [inline pkg/users/accounts-list.js:272]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221203-084140-31427c73-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/accounts-list.js [hunk 6]
```diff
@@ -183,17 +249,77 @@
...
 
+            return (aitem.title).toString().localeCompare(bitem.title.toString());
+        });
+        return direction === SortByDirection.asc ? sortedRows : sortedRows.reverse();
```
Comment (cockpituous) [inline pkg/users/accounts-list.js:276]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221203-084140-31427c73-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/delete-group-dialog.js (hunk 5 above)
Comment (cockpituous) [inline pkg/users/delete-group-dialog.js:71]: These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221203-084140-31427c73-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by garrett — DISMISSED** [review]
This is looking great! It's probably pretty close to getting to a state where we can merge it.

However, I've spotted a few issues.

Here's are screenshots for reference:

Screenshot 2022-12-05 at 09-56-48 Accounts - garrett@Bolt

Screenshot 2022-12-05 at 09-57-49 Accounts - garrett@Bolt

Issues:

- The (# more...) in the collapsed groups section should be clickable.
   - Looking at it now, should we just style it like a link-styled-button instead of a label?
   - I think this would deviate from the design a bit, but I think it probably makes more sense to look like a clickable link. It could also be implemented as a label that just looks like a link, if that makes the alignment of the text and the spacing more consistent.
- The numbers look incorrect when collapsed. They're inconsistent with the # of users when expanded. (root and demo both have users associated)
- We need dark mode colors that fit in better.
  - Dark mode group label colors are also inconsistent between groups and users, when expanded and collapsed (in different ways), as the outline and background isn't there in dark mode for non-admin groups.

Here's what it looks like in dark mode:

Screenshot 2022-12-05 at 09-59-18 Accounts - garrett@Bolt
Screenshot 2022-12-05 at 09-59-32 Accounts - garrett@Bolt

I should probably be the one to look into the dark mode colors, if PF doesn't have a way of doing this more automatically (perhaps using some color variables — possibly even from the charting set of colors — that adapt already).

File: pkg/users/accounts-list.js [hunk 7]
```diff
@@ -183,17 +249,77 @@ const mapGroupsToAccount = (accounts, groups) => {
...
+            const bitem = b.columns[idx];
+            const aname = a.columns[0];
+            const bname = b.columns[0];
 
-    const filtered_accounts = accounts.filter(account => {
-        if ((account.uid < 1000 && account.uid !== 0) ||
-                 account.shell.match(/^(\/usr)?\/sbin\/nologin/) ||
-                 account.shell === '/bin/false')
-            return false;
+            // current user is always first
```
Comment (tomasmatus) [inline pkg/users/accounts-list.js:268]: This should probably say `admin` is first?

File: pkg/users/accounts-list.js [hunk 8]
```diff
@@ -84,29 +125,54 @@ const UserActions = ({ account }) => {
...
+    if (group.isAdmin)
+        groupColorClass = "group-gold";
+    else if (group.members > 0)
+        groupColorClass = "group-cyan";
+    else
+        groupColorClass = "group-grey";
 
+    const columns = [
+        {
+            title: <Flex alignItems={{ default: 'alignItemsCenter' }}><div className={"dot " + groupColorClass} /><FlexItem>{group.name}</FlexItem></Flex>,
```
Comment (tomasmatus) [inline pkg/users/accounts-list.js:139]: You'll need to use some `sortKey` here. Right now the order when sorting by group name is not alphabetical and it changes on every second click

File: pkg/users/accounts-list.js [hunk 9]
```diff
@@ -183,17 +249,77 @@ const mapGroupsToAccount = (accounts, groups) => {
...
-                 account.shell.match(/^(\/usr)?\/sbin\/nologin/) ||
-                 account.shell === '/bin/false')
-            return false;
+            // current user is always first
+            if (admins.includes(aname.title))
+                return direction === SortByDirection.asc ? -1 : 1;
+            if (admins.includes(bname.title))
+                return direction === SortByDirection.asc ? 1 : -1;
 
+            return (aitem.title).toString().localeCompare(bitem.title.toString());
```
Comment (tomasmatus) [inline pkg/users/accounts-list.js:274]: `localeCompare` seems to be a bit problematic when sorting by id (and # of users too). What I'm seeing now is order: 1; 10; 100; 1000; 104; ... 11;

Note that this is also used in the accounts table so it's wrong in the code I wrote as well! :)

**Review by tomasmatus — CHANGES_REQUESTED** [review]
Looks great! However I found some problems when sorting rows in the group table.

File: pkg/users/accounts-list.js (hunk 9 above)
Comment (KKoukiou) [reply pkg/users/accounts-list.js]: Fixed.

File: pkg/users/accounts-list.js (hunk 8 above)
Comment (KKoukiou) [reply pkg/users/accounts-list.js:140]: Fixed

File: pkg/users/accounts-list.js (hunk 1 above)
Comment (cockpituous) [inline pkg/users/accounts-list.js:67]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221205-164019-36c705fe-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/accounts-list.js (hunk 2 above)
Comment (cockpituous) [inline pkg/users/accounts-list.js:71]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221205-164019-36c705fe-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/accounts-list.js [hunk 10]
```diff
@@ -183,17 +250,82 @@
...
+        { title: _("Accounts") },
+    ];
 
-    const accounts = mapGroupsToAccount(accountsInfo, groups);
+    const sortRows = (rows, direction, idx) => {
+        // GID and members columns are numeric
+        const isNumeric = idx == 1 || idx == 2;
+        const sortedRows = rows.sort((a, b) => {
+            const aitem = a.columns[idx].sortKey || a.columns[idx].title;
+            const bitem = b.columns[idx].sortKey || b.columns[idx].title;
```
Comment (cockpituous) [inline pkg/users/accounts-list.js:267]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221205-164019-36c705fe-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/accounts-list.js [hunk 11]
```diff
@@ -183,17 +250,82 @@
...
+            const bname = b.columns[0].sortKey;
 
-    const filtered_accounts = accounts.filter(account => {
-        if ((account.uid < 1000 && account.uid !== 0) ||
-                 account.shell.match(/^(\/usr)?\/sbin\/nologin/) ||
-                 account.shell === '/bin/false')
-            return false;
+            // administrator groups are always first
+            if (admins.includes(aname))
+                return direction === SortByDirection.asc ? -1 : 1;
```
Comment (cockpituous) [inline pkg/users/accounts-list.js:273]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221205-164019-36c705fe-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/accounts-list.js [hunk 12]
```diff
@@ -183,17 +250,82 @@
...
+            if (admins.includes(bname))
+                return direction === SortByDirection.asc ? 1 : -1;
```
Comment (cockpituous) [inline pkg/users/accounts-list.js:275]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221205-164019-36c705fe-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/accounts-list.js [hunk 13]
```diff
@@ -183,17 +250,82 @@
...
+
+            if (isNumeric)
+                return bitem - aitem;
```
Comment (cockpituous) [inline pkg/users/accounts-list.js:278]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221205-164019-36c705fe-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/accounts-list.js [hunk 14]
```diff
@@ -183,17 +250,82 @@
...
+            else
+                return aitem.localeCompare(bitem);
+        });
+        return direction === SortByDirection.asc ? sortedRows : sortedRows.reverse();
```
Comment (cockpituous) [inline pkg/users/accounts-list.js:282]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221205-164019-36c705fe-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/accounts-list.js [hunk 15]
```diff
@@ -183,17 +250,82 @@
...
+                        {groups.slice(0, 3)
+                                .map(group => {
+                                    const color = group.isAdmin ? "gold" : "cyan";
+                                    return (
+                                        <Label key={group.name} variant="filled" color={color}>
+                                            {group.name + ": " + (group.userlistPrimary.length + group.userlist.length)}
+                                        </Label>
+                                    );
+                                })}
+                        <Button key="more" className="group-more-btn" isInline variant='link' onClick={() => setIsExpanded(!isExpanded)}>
```
Comment (cockpituous) [inline pkg/users/accounts-list.js:307]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221205-164019-36c705fe-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/delete-group-dialog.js (hunk 5 above)
Comment (cockpituous) [inline pkg/users/delete-group-dialog.js:71]: These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221205-164019-36c705fe-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by tomasmatus — APPROVED** [review]

File: pkg/users/accounts-list.js (hunk 1 above)
Comment (cockpituous) [inline pkg/users/accounts-list.js:67]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221206-121249-7589670f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/accounts-list.js (hunk 2 above)
Comment (cockpituous) [inline pkg/users/accounts-list.js:71]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221206-121249-7589670f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/accounts-list.js (hunk 10 above)
Comment (cockpituous) [inline pkg/users/accounts-list.js:267]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221206-121249-7589670f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/accounts-list.js (hunk 11 above)
Comment (cockpituous) [inline pkg/users/accounts-list.js:273]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221206-121249-7589670f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/accounts-list.js (hunk 12 above)
Comment (cockpituous) [inline pkg/users/accounts-list.js:275]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221206-121249-7589670f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/accounts-list.js (hunk 13 above)
Comment (cockpituous) [inline pkg/users/accounts-list.js:278]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221206-121249-7589670f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/accounts-list.js (hunk 14 above)
Comment (cockpituous) [inline pkg/users/accounts-list.js:282]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221206-121249-7589670f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/accounts-list.js (hunk 15 above)
Comment (cockpituous) [inline pkg/users/accounts-list.js:307]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221206-121249-7589670f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/delete-group-dialog.js (hunk 5 above)
Comment (cockpituous) [inline pkg/users/delete-group-dialog.js:71]: These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17980-20221206-121249-7589670f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Pull request merged**